

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>4.1. Transcriptomics-Proteomics &mdash; SpatialFuser v1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../_static/documentation_options.js?v=5cb08e4e"></script>
      <script src="../../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="4.2. Transcriptomics-Metabolomics" href="../4.2.%20Transcriptomics-Metabolomics/RNA-Metab.html" />
    <link rel="prev" title="4. Weakly correlated multi-omics datasets across different resolution" href="../index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            SpatialFuser
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../../Installation/contents.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../../Fine%20Spatial%20Interpretation/index.html">Fine Spatial Interpretation</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../index.html">Integration and Alignment</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../1.%20Adjacent%20slices/BaristaSeq%20Alignment%20and%20Integration.html">1. Adjacent slices</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../2.%20Slices%20of%20different%20developmental%20stages/Setero-seq%20Alignment%20and%20Integration.html">2. Slices of different developmental stages</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../3.%20Spatial%20ATAC%E2%80%93RNA-seq%20dataset/Spatial%20ATAC-RNA-seq%20Integration.html">3. Spatial ATAC–RNA-seq dataset</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../index.html">4. Weakly correlated multi-omics datasets across different resolution</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">4.1. Transcriptomics-Proteomics</a></li>
<li class="toctree-l4"><a class="reference internal" href="../4.2.%20Transcriptomics-Metabolomics/RNA-Metab.html">4.2. Transcriptomics-Metabolomics</a></li>
<li class="toctree-l4"><a class="reference internal" href="../4.3.%20Proteomics-Metabolomics/Prot-Metab.html">4.3. Proteomics-Metabolomics</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../API.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">SpatialFuser</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Tutorials</a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Integration and Alignment</a></li>
          <li class="breadcrumb-item"><a href="../index.html">4. Weakly correlated multi-omics datasets across different resolution</a></li>
      <li class="breadcrumb-item active">4.1. Transcriptomics-Proteomics</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../../_sources/Tutorials/Integration and Alignment/4. Weakly correlated multi-omics datasets across different resolution/4.1. Transcriptomics-Proteomics/RNA-Prot.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="4.1.-Transcriptomics-Proteomics">
<h1>4.1. Transcriptomics-Proteomics<a class="headerlink" href="#4.1.-Transcriptomics-Proteomics" title="Link to this heading"></a></h1>
<p>Pairwise integration and alignment were performed on the multi-omics mouse cerebellum dataset, specifically between the PLATO spatial proteomics and MALDI-MSI spatial metabolomics consecutive slices.</p>
<p>The MAGIC-seq slice captured 1,677 spatial spots at a resolution of 32 μm, profiling 16,116 genes. The PLATO slice provides high-throughput proteomic profiling aligned with the transcriptomic slice, identifying 5,722 protein groups after AI enhancement and quality control.</p>
<section id="Preparation">
<h2>Preparation<a class="headerlink" href="#Preparation" title="Link to this heading"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">time</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">spatialFuser</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;..&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/louvain/__init__.py:54: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools&lt;81.
  from pkg_resources import get_distribution, DistributionNotFound
</pre></div></div>
</div>
</section>
<section id="Hyper-Parameters-setting">
<h2>Hyper-Parameters setting<a class="headerlink" href="#Hyper-Parameters-setting" title="Link to this heading"></a></h2>
<p>All hyperparameters are stored in the variable args. To display them during initialization, use the function call <em>args = def_training_args(show_detail=True)</em>.</p>
<p>For the high-throughput MAGIC-seq slice and PLATO slice, two-layer MCGATE models was employed to learn embeddings.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load args:</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;============================================&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=              Setting Params              =&quot;</span><span class="p">)</span>
<span class="n">slice_1_args</span> <span class="o">=</span> <span class="n">def_training_args</span><span class="p">()</span>
<span class="n">slice_1_args</span><span class="o">.</span><span class="n">hidden</span> <span class="o">=</span> <span class="p">[</span><span class="mi">512</span><span class="p">,</span> <span class="mi">32</span><span class="p">]</span>
<span class="n">slice_1_args</span><span class="o">.</span><span class="n">epochs</span> <span class="o">=</span> <span class="mi">500</span>
<span class="n">slice_1_args</span><span class="o">.</span><span class="n">lr</span> <span class="o">=</span> <span class="mf">3e-3</span>
<span class="n">slice_1_args</span><span class="o">.</span><span class="n">K</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">slice_1_args</span><span class="o">.</span><span class="n">heads</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">slice_1_args</span><span class="o">.</span><span class="n">alpha</span> <span class="o">=</span> <span class="mi">0</span>

<span class="n">slice_2_args</span> <span class="o">=</span> <span class="n">def_training_args</span><span class="p">()</span>
<span class="n">slice_2_args</span><span class="o">.</span><span class="n">hidden</span> <span class="o">=</span> <span class="p">[</span><span class="mi">512</span><span class="p">,</span> <span class="mi">32</span><span class="p">]</span>
<span class="n">slice_2_args</span><span class="o">.</span><span class="n">epochs</span> <span class="o">=</span> <span class="mi">500</span>
<span class="n">slice_2_args</span><span class="o">.</span><span class="n">lr</span> <span class="o">=</span> <span class="mf">1e-3</span>
<span class="n">slice_2_args</span><span class="o">.</span><span class="n">K</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">slice_2_args</span><span class="o">.</span><span class="n">heads</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">slice_2_args</span><span class="o">.</span><span class="n">alpha</span> <span class="o">=</span> <span class="mi">0</span>

<span class="n">integration_args</span> <span class="o">=</span> <span class="n">def_training_args</span><span class="p">()</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">hidden</span> <span class="o">=</span> <span class="p">[</span><span class="mi">32</span><span class="p">,</span> <span class="mi">32</span><span class="p">]</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">fusion_epoch</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">lr</span> <span class="o">=</span> <span class="mf">3e-3</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">match_step_size</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">tau</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">roi_radius</span> <span class="o">=</span> <span class="mf">0.02</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">epsilon</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">m_top_K</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">beta_rec</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">beta_dir</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="n">integration_args</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">False</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
============================================
=              Setting Params              =
</pre></div></div>
</div>
</section>
<section id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Link to this heading"></a></h2>
<p>SpatialFuser provides a built-in data loading and preprocessing module, <em>SpatialFuserDataLoader</em>. The required inputs include hyperparameters, data_dir (the dataset storage directory), data_tech (either “seq-based” or “image-based”), and files (a list of h5ad files to be loaded).</p>
<p>For spatial omics data, <em>SpatialFuserDataLoader</em> constructs a KNN adjacency graph based on the specified value of K to support graph neural network training.</p>
<p>For seq-based data, spatially variable genes are extracted according to n_svgs to simplify the model.</p>
<p>All AnnData objects are normalized, log-transformed, and subsequently converted into PyG objects for model input.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load data:</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;============================================&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=               Loading Data               =&quot;</span><span class="p">)</span>
<span class="n">slice_1_dataLoader</span> <span class="o">=</span> <span class="n">SpatialFuserDataLoader</span><span class="p">(</span><span class="n">slice_1_args</span><span class="p">,</span>
                                            <span class="n">data_dir</span><span class="o">=</span><span class="s1">&#39;/public8/lilab/student/whcai/Integration/data/cell_fangqing_zhao&#39;</span><span class="p">,</span>
                                            <span class="n">data_tech</span><span class="o">=</span><span class="s1">&#39;seq-based&#39;</span><span class="p">,</span>
                                            <span class="n">files</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cerebellum-MAGIC-seq_raw.h5ad&#39;</span><span class="p">])</span>
<span class="n">slice_1_dataLoader</span><span class="o">.</span><span class="n">load_adata</span><span class="p">()</span>
<span class="n">slice_1_dataLoader</span><span class="o">.</span><span class="n">pre_processing</span><span class="p">(</span><span class="n">n_svgs</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">k_cutoff</span><span class="o">=</span><span class="n">slice_1_args</span><span class="o">.</span><span class="n">K</span><span class="p">,</span> <span class="n">batch_label</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">slice_1_dataLoader</span><span class="o">.</span><span class="n">generate_minibatch</span><span class="p">(</span><span class="n">loader_type</span><span class="o">=</span><span class="s1">&#39;RandomNodeLoader&#39;</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

<span class="n">slice_2_dataLoader</span> <span class="o">=</span> <span class="n">SpatialFuserDataLoader</span><span class="p">(</span><span class="n">slice_2_args</span><span class="p">,</span>
                                            <span class="n">data_dir</span><span class="o">=</span><span class="s1">&#39;/public8/lilab/student/whcai/Integration/data/cell_fangqing_zhao&#39;</span><span class="p">,</span>
                                            <span class="n">data_tech</span><span class="o">=</span><span class="s1">&#39;seq-based&#39;</span><span class="p">,</span>
                                            <span class="n">files</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cerebellum-PLATO_raw.h5ad&#39;</span><span class="p">])</span>
<span class="n">slice_2_dataLoader</span><span class="o">.</span><span class="n">load_adata</span><span class="p">()</span>
<span class="n">slice_2_dataLoader</span><span class="o">.</span><span class="n">pre_processing</span><span class="p">(</span><span class="n">n_svgs</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">k_cutoff</span><span class="o">=</span><span class="n">slice_2_args</span><span class="o">.</span><span class="n">K</span><span class="p">,</span> <span class="n">batch_label</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="n">slice_2_dataLoader</span><span class="o">.</span><span class="n">generate_minibatch</span><span class="p">(</span><span class="n">loader_type</span><span class="o">=</span><span class="s1">&#39;RandomNodeLoader&#39;</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

<span class="n">slice_1_dataLoader</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Region&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">slice_1_dataLoader</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cluster&#39;</span><span class="p">]</span>
<span class="n">slice_2_dataLoader</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Region&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">slice_2_dataLoader</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cluster&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
============================================
=               Loading Data               =
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:61: UserWarning: `flavor=&#39;seurat_v3&#39;` expects raw count data, but non-integers were found.
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: adata.X seems to be already log-transformed.
=         Calculating spatial graph        =
The PyG data u create is qualified
=The graph contains 8385 edges, 1677 cells=
=   5.0000 neighbors per cell on average   =
=              subgraph Info             =
============================================
=           Batch 0: 1677 nodes           =
=   4.0000 neighbors per cell on average   =
batch:[1.], node num:[1677]
============================================
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:61: UserWarning: `flavor=&#39;seurat_v3&#39;` expects raw count data, but non-integers were found.
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: adata.X seems to be already log-transformed.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/scanpy/preprocessing/_normalization.py:196: UserWarning: Some cells have zero counts
  warn(UserWarning(&#39;Some cells have zero counts&#39;))
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
=         Calculating spatial graph        =
The PyG data u create is qualified
=The graph contains 8385 edges, 1677 cells=
=   5.0000 neighbors per cell on average   =
=              subgraph Info             =
============================================
=           Batch 0: 1677 nodes           =
=   4.0000 neighbors per cell on average   =
batch:[2.], node num:[1677]
============================================
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_19184/2482162417.py:20: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
  slice_1_dataLoader.adata.obs[&#39;Region&#39;] = slice_1_dataLoader.adata.obs[&#39;cluster&#39;]
/tmp/ipykernel_19184/2482162417.py:21: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
  slice_2_dataLoader.adata.obs[&#39;Region&#39;] = slice_2_dataLoader.adata.obs[&#39;cluster&#39;]
</pre></div></div>
</div>
</section>
<section id="Train">
<h2>Train<a class="headerlink" href="#Train" title="Link to this heading"></a></h2>
<p>The MCGATEs are first pre-trained individually, after which the fusion layer and matching layer are incorporated into the training process.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># train</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;============================================&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=              Begin to Train              =&quot;</span><span class="p">)</span>
<span class="n">training_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="p">[</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">],</span> <span class="n">trainer</span> <span class="o">=</span> <span class="n">train_integration</span><span class="p">([</span><span class="n">slice_1_args</span><span class="p">,</span> <span class="n">slice_2_args</span><span class="p">,</span> <span class="n">integration_args</span><span class="p">],</span>
                                              <span class="p">[</span><span class="n">slice_1_dataLoader</span><span class="p">,</span> <span class="n">slice_2_dataLoader</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=            Training Finished!            =&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total time elapsed: </span><span class="si">{:.4f}</span><span class="s2">s&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">training_time</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;============================================&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
============================================
=              Begin to Train              =
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/public8/lilab/student/whcai/Integration/model/SpatialFuser/spatialFuser/train.py:137: UserWarning: torch.sparse.SparseTensor(indices, values, shape, *, device=) is deprecated.  Please use torch.sparse_coo_tensor(indices, values, shape, dtype=, device=). (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:605.)
  self.adata1_adata2atial_adj = torch.sparse.FloatTensor(batch_in_data1.edge_index,
Epoch 500 || adata1_pretrain_loss: 0.128 || adata2_pretrain_loss: 0.053 || : 100%|████| 500/500 [00:29&lt;00:00, 17.16it/s]
Epoch 200 || Fusion_loss : 0.1595 || MSE_loss: 0.1882 || Dir_loss: 0.4295 || : 100%|██| 200/200 [00:12&lt;00:00, 15.88it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
=            Training Finished!            =
Total time elapsed: 43.1083s
============================================
</pre></div></div>
</div>
</section>
<section id="Spatial-domain-detection-after-integration">
<h2>Spatial domain detection after integration<a class="headerlink" href="#Spatial-domain-detection-after-integration" title="Link to this heading"></a></h2>
<p>SpatialFuser provides an evaluation module, <em>metrics</em>, for assessing tissue domain detection tasks. It treats the Region column in anndata.obs as the ground truth and, based on the provided embed_label (an array stored in anndata.obsm), automatically computes five metrics (ARI, AMI, Homogeneity, Completeness, and V-Measure) under clustering methods including Leiden, Louvain, and Mclust.</p>
<p>Here, we only present the spatial domains and Umap colored by Louvain.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># spatial domain detection</span>
<span class="n">adata1_leiden_result</span><span class="p">,</span> <span class="n">adata1_louvain_result</span><span class="p">,</span> <span class="n">adata1_mclust_result</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span>
                                                                            <span class="n">save_loc</span><span class="o">=</span><span class="s1">&#39;_slice1.png&#39;</span><span class="p">,</span>
                                                                            <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                                                            <span class="n">resolution</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                                                                            <span class="n">spot_size</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span>
                                                                            <span class="n">cluster_label</span><span class="o">=</span><span class="s1">&#39;Region&#39;</span><span class="p">,</span>
                                                                            <span class="n">plot_color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="p">],</span>
                                                                            <span class="n">mclust_model</span><span class="o">=</span><span class="s1">&#39;EEE&#39;</span><span class="p">,</span>
                                                                            <span class="n">embed_label</span><span class="o">=</span><span class="s1">&#39;fused_embedding&#39;</span><span class="p">,</span>
                                                                            <span class="n">vis</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                                                            <span class="n">save</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">adata2_leiden_result</span><span class="p">,</span> <span class="n">adata2_louvain_result</span><span class="p">,</span> <span class="n">adata2_mclust_result</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">(</span><span class="n">adata2</span><span class="p">,</span>
                                                                            <span class="n">save_loc</span><span class="o">=</span><span class="s1">&#39;_slice2.png&#39;</span><span class="p">,</span>
                                                                            <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
                                                                            <span class="n">resolution</span><span class="o">=</span><span class="mf">0.14</span><span class="p">,</span>
                                                                            <span class="n">spot_size</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span>
                                                                            <span class="n">cluster_label</span><span class="o">=</span><span class="s1">&#39;Region&#39;</span><span class="p">,</span>
                                                                            <span class="n">plot_color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;louvain&quot;</span><span class="p">],</span>
                                                                            <span class="n">mclust_model</span><span class="o">=</span><span class="s1">&#39;EEE&#39;</span><span class="p">,</span>
                                                                            <span class="n">embed_label</span><span class="o">=</span><span class="s1">&#39;fused_embedding&#39;</span><span class="p">,</span>
                                                                            <span class="n">vis</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                                                            <span class="n">save</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">region_color</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Fiber tracts&#39;</span><span class="p">:</span> <span class="s2">&quot;#a03b32&quot;</span><span class="p">,</span>
                <span class="s1">&#39;Granular layer&#39;</span><span class="p">:</span> <span class="s2">&quot;#ffcf4f&quot;</span><span class="p">,</span>
                <span class="s1">&#39;Lateral recess&#39;</span><span class="p">:</span> <span class="s2">&quot;#33658A&quot;</span><span class="p">,</span>
                <span class="s1">&#39;Molecular layer&#39;</span><span class="p">:</span> <span class="s2">&quot;#878bb4&quot;</span><span class="p">,</span>

                <span class="s1">&#39;0&#39;</span><span class="p">:</span> <span class="s2">&quot;#878bb4&quot;</span><span class="p">,</span>
                <span class="s1">&#39;1&#39;</span><span class="p">:</span> <span class="s2">&quot;#ffcf4f&quot;</span><span class="p">,</span>
                <span class="s1">&#39;2&#39;</span><span class="p">:</span> <span class="s2">&quot;#a03b32&quot;</span><span class="p">,</span>
                <span class="s1">&#39;3&#39;</span><span class="p">:</span> <span class="s2">&quot;#33658A&quot;</span><span class="p">,</span>
                <span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;#89606a&quot;</span><span class="p">,</span>
                <span class="mi">2</span><span class="p">:</span> <span class="s2">&quot;#71a2b6&quot;</span><span class="p">,</span>
                <span class="p">}</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span>
              <span class="n">img_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
              <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Region&quot;</span><span class="p">,</span> <span class="s2">&quot;louvain&quot;</span><span class="p">],</span>
              <span class="n">spot_size</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span>
              <span class="n">title</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Original Annotation&#39;</span><span class="p">,</span> <span class="s1">&#39;Louvain ARI: </span><span class="si">{:.2f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">adata1_louvain_result</span><span class="p">[</span><span class="s1">&#39;ARI&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])],</span>
              <span class="n">wspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
              <span class="n">palette</span><span class="o">=</span><span class="n">region_color</span><span class="p">,</span>
              <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
              <span class="n">save</span><span class="o">=</span><span class="s1">&#39;_RNA_Prot_RNA.png&#39;</span>
              <span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">adata2</span><span class="p">,</span>
              <span class="n">img_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
              <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Region&quot;</span><span class="p">,</span> <span class="s2">&quot;louvain&quot;</span><span class="p">],</span>
              <span class="n">spot_size</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span>
              <span class="n">title</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Original Annotation&#39;</span><span class="p">,</span> <span class="s1">&#39;Louvain ARI: </span><span class="si">{:.2f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">adata2_louvain_result</span><span class="p">[</span><span class="s1">&#39;ARI&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])],</span>
              <span class="n">wspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
              <span class="n">palette</span><span class="o">=</span><span class="n">region_color</span><span class="p">,</span>
              <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
              <span class="n">save</span><span class="o">=</span><span class="s1">&#39;_RNA_Prot_Prot.png&#39;</span>
              <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2025-08-25 21:01:18.555406: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-08-25 21:01:18.757985: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-08-25 21:01:20.326164: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library &#39;libnvinfer.so.7&#39;; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/cv2/../../lib64::/usr/local/cuda-12.4/lib64:/public8/lilab/student/whcai/myR/R-4.1.2/lib/
2025-08-25 21:01:20.326483: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library &#39;libnvinfer_plugin.so.7&#39;; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/cv2/../../lib64::/usr/local/cuda-12.4/lib64:/public8/lilab/student/whcai/myR/R-4.1.2/lib/
2025-08-25 21:01:20.326510: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
R[write to console]: Package &#39;mclust&#39; version 6.0.1
Type &#39;citation(&#34;mclust&#34;)&#39; for citing this R package in publications.

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
fitting ...
  |======================================================================| 100%
fitting ...
  |======================================================================| 100%
WARNING: saving figure to file figures/show_RNA_Prot_RNA.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_9_2.png" src="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_9_2.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/show_RNA_Prot_Prot.png
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_9_4.png" src="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_9_4.png" />
</div>
</div>
</section>
<section id="Modality-bias-correction">
<h2>Modality bias correction<a class="headerlink" href="#Modality-bias-correction" title="Link to this heading"></a></h2>
<p>UMAP visualizations colored by ground truth annotation (left) and modalities (right), illustrating that the model preserves true biological differences while achieving effective integration.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># modality bias correction (integration)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">adata1</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;fused_embedding&#39;</span><span class="p">],</span> <span class="n">adata2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;fused_embedding&#39;</span><span class="p">]]))</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">adata1</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">],</span> <span class="n">adata2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]])</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata1</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">adata2</span><span class="o">.</span><span class="n">obs</span><span class="p">])</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batches&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
           <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Region&quot;</span><span class="p">,</span> <span class="s2">&quot;batches&quot;</span><span class="p">],</span>
           <span class="n">wspace</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
           <span class="n">palette</span><span class="o">=</span><span class="n">region_color</span><span class="p">,</span>
           <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
           <span class="n">save</span><span class="o">=</span><span class="s1">&#39;_RNA_Prot_check_batch.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap_RNA_Prot_check_batch.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/whcai/anaconda3/envs/PyG/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_11_2.png" src="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_11_2.png" />
</div>
</div>
</section>
<section id="Show-alignment">
<h2>Show alignment<a class="headerlink" href="#Show-alignment" title="Link to this heading"></a></h2>
<p>Alignment across slices, colored by region labels (300 alignment pairs shown for clarity)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># show alignment</span>
<span class="n">adata1_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                          <span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="n">adata1</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span>
                          <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="n">adata1</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span>
                          <span class="s1">&#39;celltype&#39;</span><span class="p">:</span> <span class="n">adata1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Region&#39;</span><span class="p">]})</span>
<span class="n">adata2_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata2</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                          <span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="n">adata2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span>
                          <span class="s1">&#39;y&#39;</span><span class="p">:</span> <span class="n">adata2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span>
                          <span class="s1">&#39;celltype&#39;</span><span class="p">:</span> <span class="n">adata2</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Region&#39;</span><span class="p">]})</span>
<span class="n">matching</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">trainer</span><span class="o">.</span><span class="n">match_in_adata1</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">(),</span> <span class="n">trainer</span><span class="o">.</span><span class="n">match_in_adata2</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()])</span>
<span class="n">multi_align</span> <span class="o">=</span> <span class="n">match_3D_multi</span><span class="p">(</span><span class="n">adata1_df</span><span class="p">,</span> <span class="n">adata2_df</span><span class="p">,</span> <span class="n">matching</span><span class="p">,</span> <span class="n">meta</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span>
                             <span class="n">scale_coordinate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">subsample_size</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">exchange_xy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">multi_align</span><span class="o">.</span><span class="n">draw_3D</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="s1">&#39;all_type&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="n">line_width</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">],</span> <span class="n">line_color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span>
                    <span class="n">hide_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_error</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">only_show_correct</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">only_show_error</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="n">line_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dataset1: 4 cell types; dataset2: 4 cell types;
                    Total :4 celltypes; Overlap: 4 cell types
                    Not overlap :[[]]
Subsampled 300 pairs from 1679
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_13_1.png" src="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_13_1.png" />
</div>
</div>
</section>
<section id="Spots-mapping-accuracy">
<h2>Spots mapping accuracy<a class="headerlink" href="#Spots-mapping-accuracy" title="Link to this heading"></a></h2>
<p>The all_matching function returns two plots:</p>
<ol class="arabic simple">
<li><p>A Sankey plot illustrating spots type correspondence based on alignments from SpatialFuser.</p></li>
<li><p>A histogram showing the similarity score distributions from random matching (bottom) and SpatialFuser matching (top) in pairwise alignments.</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># evaluate alignment</span>
<span class="n">valid_ratio</span><span class="p">,</span> <span class="n">accuracy</span> <span class="o">=</span> <span class="n">all_matching</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">,</span>
                                     <span class="mf">0.95</span><span class="p">,</span>
                                     <span class="mf">0.02</span><span class="p">,</span>
                                     <span class="n">save_loc</span><span class="o">=</span><span class="s1">&#39;./figures/&#39;</span><span class="p">,</span>
                                     <span class="n">file_name</span><span class="o">=</span><span class="s1">&#39;RNA_Prot&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

QC threshold：0.95%  -&gt; 0.9977
valid matching ratio：99.1% (1662/1677)
spots mapping acc：84.06% (1397/1662)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_15_1.png" src="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_15_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_15_2.png" src="../../../../_images/Tutorials_Integration_and_Alignment_4._Weakly_correlated_multi-omics_datasets_across_different_resolution_4.1._Transcriptomics-Proteomics_RNA-Prot_15_2.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../index.html" class="btn btn-neutral float-left" title="4. Weakly correlated multi-omics datasets across different resolution" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../4.2.%20Transcriptomics-Metabolomics/RNA-Metab.html" class="btn btn-neutral float-right" title="4.2. Transcriptomics-Metabolomics" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Wenhao Cai.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>